trackintel/preprocessing/positionfixes.py

```diff
@@ -173,187 +177,335 @@ def generate_triplegs(positionfixes, staypoints=None, *args, **kwargs):
+                                        geometry='geom', crs=ret_pfs.crs)
+        else:
+            ret_tpls = gpd.GeoDataFrame(generated_triplegs, geometry='geom', crs=ret_pfs.crs)
 
-                started_at = pfs_tripleg['tracked_at'].iloc[0]
-                finished_at = pfs_tripleg['tracked_at'].iloc[-1]
+        ret_tpls.index.name = 'id'
 
-                coords = list(pfs_tripleg.geometry.apply(lambda r: (r.x, r.y)))
+        # assign tripleg_id to positionfixes
+        ret_pfs['tripleg_id'] = np.nan
 
-                if len(coords) > 1:
-                    generated_triplegs.append({
-                        'id': curr_tripleg_id,
-                        'user_id': user_id_this,
-                        'started_at': started_at,  # pfs_tripleg['tracked_at'].iloc[0],
-                        'finished_at': finished_at,  # pfs_tripleg['tracked_at'].iloc[-1],
-                        'geom': LineString(coords)
-                    })
-                    curr_tripleg_id += 1
+        for tripleg_id, tripleg in ret_tpls.iterrows():
+            pfs_ids = tripleg['pfs_ids']
+            ret_pfs.loc[pfs_ids, 'tripleg_id'] = tripleg_id
+        ret_tpls = ret_tpls.drop('pfs_ids', axis=1)
 
-        # Case 2: Staypoints exist but there is no user_id given
-        # TODO Not so efficient, always matching on the time (as things are sorted anyways).
-        elif staypoints is not None:
-            stps = staypoints.loc[staypoints['user_id'] == user_id_this].sort_values('started_at')
-            stps = stps.to_dict('records')
-            for stp1, stp2 in zip(list(stps), list(stps)[1:]):
-                # Get all positionfixes that lie between these two staypoints.
-                pfs_tripleg = pfs[(stp1['finished_at'] <= pfs['tracked_at']) & \
-                                  (pfs['tracked_at'] <= stp2['started_at'])].sort_values('tracked_at')
-
-                coords = list(pfs_tripleg.geometry.apply(lambda r: (r.x, r.y)))
-                if len(coords) > 1:
-                    generated_triplegs.append({
-                        'id': curr_tripleg_id,
-                        'user_id': user_id_this,
-                        'started_at': pfs_tripleg['tracked_at'].iloc[0],
-                        'finished_at': pfs_tripleg['tracked_at'].iloc[-1],
-                        'geom': LineString(list(pfs_tripleg.geometry.apply(lambda r: (r.x, r.y))))
-                    })
-                    curr_tripleg_id += 1
-
-        # case 3: Only positionfixes with staypoint id for tripleg generation
+        ## ensure dtype consistency
+        # tpls id (generated by this function) should be float (missing values)
+        ret_pfs['tripleg_id'] = ret_pfs['tripleg_id'].astype('float')
+        # user_id of tpls should be the same as ret_pfs
+        ret_tpls['user_id'] = ret_tpls['user_id'].astype(ret_pfs['user_id'].dtype)
+
+        # todo: proposed fix: ret_triplegs = ret_triplegs.set_index('id')
```

Comment:
ret_triplegs already has the id in the index? line 242

Reply:
Ah no, actually not. I fixed it 👍